File: src/main/java/org/dukcode/ps/codetree/trail02/chapter02/lesson01/TestStarOutputWithRecursiveFunction2.java
package org.dukcode.ps.codetree.trail02.chapter02.lesson01;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/test-star-output-with-recursive-function-2/description">LINK</a>
 */
public class TestStarOutputWithRecursiveFunction2 {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int n = Integer.parseInt(br.readLine());
    printStars(n);

    br.close();
    bw.close();

  }

  private static void printStars(int n) throws IOException {
    if (n == 0) {
      return;
    }

    printLine(n);
    printStars(n - 1);
    printLine(n);
  }

  private static void printLine(int n) throws IOException {
    for (int i = 0; i < n; ++i) {
      bw.write('*');
      bw.write(' ');
    }
    bw.newLine();
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter01/lesson05/TestStarOutputWithRecursiveFunction2.kt
package orgMain.ps.codetree.trail02.chapter01.lesson05.testStarOutputWithRecursiveFunction2

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/test-star-output-with-recursive-function-2/description">LINK</a>
 */
fun main() {
    val n = br.readLine().toInt()

    printStars(n)

    bw.flush()
}

private fun printStars(n: Int) {
    if (n == 0) {
        return
    }

    bw.write("* ".repeat(n))
    bw.newLine()

    printStars(n - 1)

    bw.write("* ".repeat(n))
    bw.newLine()
}
